### PR TITLE
[KYUUBI #2285] trino's result fetching method is changed to a streaming iterator mode to avoid hold data at server side

### PR DESCRIPTION
**Pull request by iodone**
### _Why are the changes needed?_

1. Change to a true iterator implementation to get the results of Trino Server to avoid caching data in memory.
2. Supports FETCH_NEXT mode only, consistent with Trino Server's capabilities

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by pan3793** [comment]
> Supports FETCH_NEXT mode only, consistent with Trino Server's capabilities

IIRC, it may break HUE export data functionality. Can we support both?

**Comment by iodone** [comment]
> > Supports FETCH_NEXT mode only, consistent with Trino Server's capabilities
> 
> IIRC, it may break HUE export data functionality. Can we support both?

IMO, trino is designed to return results based on the pipeline model and does not have the capability of other fetch order types. If we want to support it, need to cache the trino results on the Server side, but Kyuubi Server is not suitable for caching results as the Gateway, because it may lead to OOM.

image
https://trino.io/docs/current/sql/select.html?highlight=fetch#limit-or-fetch-first-clause 

It was found that Trino SQL originally supported these order modes.

**Comment by pan3793** [comment]
I think we can give a option to cache the result on engine side, it will not affect Kyuubi server even if engine crash by OOM. Another idea is re-run query if the client request the fetch FIRST after consuming, it's usually for export cases.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/operation/FetchIterator.scala [hunk 1]
```diff
@@ -109,3 +109,26 @@ class IterableFetchIterator[A](iterable: Iterable[A]) extends FetchIterator[A] {
     }
   }
 }
+
+object FetchIterator {
+
+  def from[A](iter: Iterator[A]): FetchIterator[A] = new FetchIterator[A] {
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/operation/FetchIterator.scala:115]: is this so common to have this name?

**Review by yaooqinn — APPROVED** [review]
LGTM, only one comment

File: kyuubi-common/src/main/scala/org/apache/kyuubi/operation/FetchIterator.scala (hunk 1 above)
Comment (iodone) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/operation/FetchIterator.scala]: OK, I modified it to `fromIterator`, just like the `Future.fromTry` API

Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/operation/FetchIterator.scala]: shouldn't it be as same as IterableFetchIterator?

**Comment by hddong** [comment]
> > I think we can give an option to cache the result on engine side, it will not affect Kyuubi server even if engine crash by OOM. Another idea is re-run query if the client request the fetch FIRST after consuming, it's usually for export cases.
> 
> Can we consider implementing FETCH_FIRST in KyuybiJDBC Client, so that keep the immediate result return feature of trino?

Or we just Keep the previous `execute(): Iterable[List[Any]]` as `executeCollect(): Iterable[List[Any]]` , and run:
```
iter =
    if (incrementalCollect) {
      info("Execute in incremental collect mode")
      FetchIterator.fromIterator(trinoStatement.execute())
    } else {
      info("Execute in full collect mode")
      new ArrayFetchIterator(trinoStatement.executeCollect())
    }
```

**Comment by pan3793** [comment]
> Can we consider implementing FETCH_FIRST in KyuubiJDBC Client, so that keep the immediate result return feature of trino?

There are many language client for Kyuubi, I prefer to implement this feature in engine, like Spark full collect mode and incremental collect mode

**Comment by iodone** [comment]
> > > I think we can give an option to cache the result on engine side, it will not affect Kyuubi server even if engine crash by OOM. Another idea is re-run query if the client request the fetch FIRST after consuming, it's usually for export cases.
> > 
> > 
> > Can we consider implementing FETCH_FIRST in KyuybiJDBC Client, so that keep the immediate result return feature of trino?
> 
> Or we just Keep the previous `execute(): Iterable[List[Any]]` as `executeCollect(): Iterable[List[Any]]` , and run:
> 
> ```
> iter =
>     if (incrementalCollect) {
>       info("Execute in incremental collect mode")
>       FetchIterator.fromIterator(trinoStatement.execute())
>     } else {
>       info("Execute in full collect mode")
>       new ArrayFetchIterator(trinoStatement.executeCollect())
>     }
> ```

Good advice, I'll try it, thanks

File: externals/kyuubi-trino-engine/src/main/scala/org/apache/kyuubi/engine/trino/TrinoStatement.scala [hunk 2]
```diff
@@ -80,63 +74,34 @@ class TrinoStatement(
...
-        if (data != null) {
-          data.asScala.map(_.asScala.toList)
-            .foreach(e => putOrThrow(rowQueue, e))
+  def execute(): Iterator[List[Any]] = {
+    Iterator.continually {
+      @tailrec
+      def getData(): (Boolean, List[List[Any]]) = {
+        if (trino.isRunning) {
+          val data = trino.currentData().getData()
+          trino.advance()
```
[the remark this answers is not in the document]
Comment (iodone) [reply externals/kyuubi-trino-engine/src/main/scala/org/apache/kyuubi/engine/trino/TrinoStatement.scala:83]: `trino.advance` must be called each time to ensure that the results obtained each time is the latest, if put into 87line will lead to keep getting the same data

**Review by hddong — APPROVED** [review]

File: externals/kyuubi-trino-engine/src/test/scala/org/apache/kyuubi/engine/trino/TrinoStatementSuite.scala [hunk 3]
```diff
@@ -52,16 +52,16 @@ class TrinoStatementSuite extends WithTrinoContainerServer {
       assert(this.schema === trinoStatement.getCurrentDatabase)
 
       val trinoStatement2 = TrinoStatement(trinoContext, kyuubiConf, "use sf1")
-      trinoStatement2.execute()
-
+      // if trinoStatement.execute return iterator is lazy, call toArray to strict evaluation
```
Comment (SteNicholas) [inline externals/kyuubi-trino-engine/src/test/scala/org/apache/kyuubi/engine/trino/TrinoStatementSuite.scala:55]: How does the end user knows whether `trinoStatement.execute` is lazy?

File: externals/kyuubi-trino-engine/src/main/scala/org/apache/kyuubi/engine/trino/TrinoStatement.scala [hunk 4]
```diff
@@ -80,63 +74,34 @@ class TrinoStatement(
...
+            (true, data.asScala.toList.map(_.asScala.toList))
+          } else {
+            getData()
+          }
+        } else {
+          Verify.verify(trino.isFinished)
+          val finalStatus = trino.finalStatusInfo()
+          if (finalStatus.getError() != null) {
+            throw KyuubiSQLException(
+              s"Query ${finalStatus.getId} failed: ${finalStatus.getError.getMessage}")
```
Comment (SteNicholas) [inline externals/kyuubi-trino-engine/src/main/scala/org/apache/kyuubi/engine/trino/TrinoStatement.scala:94]: If the user invokes the `getQueryId` interface, is this query id same as the returned id of `getQueryId`?

Comment (iodone) [reply externals/kyuubi-trino-engine/src/main/scala/org/apache/kyuubi/engine/trino/TrinoStatement.scala:94]: yes，the `queryId` is generated by trinoServer at the time of the first query and is unique throughout the query process

**Comment by yaooqinn** [comment]
Hi @hddong, ready to merge?

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**